simple-backend/nlpviewer_backend/handlers/document.py

```diff
@@ -65,23 +128,92 @@ def edit_ontology(request, document_id):
 
 @require_login
 def query(request, document_id):
+    """Retrieves a document by id.
+    
+    Retrieves a document by id.
+    The function is accessible for users with 'view' or 'editor' permission of the project and the owner of the project.
+
+    Args:
+        document_id:
+            The id of the document.
+
+    Returns:
+        A json response of the modified document or forbidden or not found.
+        example:
+
+            {"id": 42, "name": "project1-doc1-example", "project_id": 5, "textPack": "...", ...}    
+    """
+    # if doc doen't exist
+    try:
+        doc = Document.objects.get(pk=document_id)
+    except ObjectDoesNotExist: 
+        return HttpResponse('not found', status=404)
+
+    # check permissions
+    if (not request.user.has_perm('nlpviewer_backend.viewer_project', doc.project) and
```

Comment:
The permission check has been called many times. I have a suggestion for this, you can create a decorator function for this check `doc_permission(permission_name: str)`, when using this on a function, all you need to do is to put `@permission(some_permission_name)` before each function. For example, in this case, you will simply write `@permission('nlpviewer_backend.viewer_project')`. Now you can let your decorator function to access the parameter of this function, say the `doc_id`, then check if the `doc_id` exists, and check if the permission is OK. You can take a look here for example:

https://stackoverflow.com/questions/1010080/python-get-original-function-arguments-in-decorator

You can also follow the existing decorator such as `require_login`